core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -774,7 +817,14 @@ private[log] class Cleaner(val id: Int,
        *   2) The message doesn't has value but it can't be deleted now.
        */
       val latestOffsetForKey = record.offset() >= foundOffset
-      val isRetainedValue = record.hasValue || retainDeletes
+      val isLatestVersion = batch.magic() >= RecordBatch.MAGIC_VALUE_V2
+      var shouldRetainDeletes = true
+      if (isLatestVersion)
+        shouldRetainDeletes = (batch.deleteHorizonSet() && currentTime < batch.deleteHorizonMs()) || 
+                              (!batch.deleteHorizonSet() && currentTime < newBatchDeleteHorizonMs)
```

Comment:
Hmm, if deleteHorizonSet is not set, we shouldn't be deleting the tombstone. So, not sure what newBatchDeleteHorizonMs is intended for.

Reply:
Oh, this is used as a means to help the tests in LogCleanerTest.scala pass. LogCleanerTest usually wants the tombstones removed in a single pass (but that pass is usually used for setting the delete horizon ms, which means without doing the above, we would be unable to remove tombstones). Therefore, by adding the ```newBatchDeleteHorizonMs``` argument (which is passed in by MemoryRecords), whenever LogCleaner calls clean log with the current time marked as ```Log.MAX_VALUE```, we will be able to remove the tombstones / control records in one pass.